qutip/solver/mcsolve.py

```diff
@@ -0,0 +1,311 @@
+        trajectories until the error on the expectation values is lower than
+        this tolerance. The error is computed using jackknife resampling.
+        ``target_tol`` can be an absolute tolerance or a pair of absolute and
+        relative tolerance, in that order. Lastly, it can be a list of pairs of
+        (atol, rtol) for each e_ops.
+
+    timeout : float [optional]
+        Maximum time for the evolution in second. When reached, no more
+        trajectories will be computed. Overwrite the option of the same name.
+
+    Returns
+    -------
+    results : :class:`qutip.solver.Result`
+        Object storing all results from the simulation. Which results is saved
+        depends on the presence of ``e_ops`` and the options used. ``collapse``
+        and ``photocurrent`` is available to Monte Carlo simulation results.
+    """
+    H = QobjEvo(H, args=args, tlist=tlist)
+    if not isinstance(c_ops, (list, tuple)):
+        c_ops = [c_ops]
+    c_ops = [QobjEvo(c_op, args=args, tlist=tlist) for c_op in c_ops]
+
+    if len(c_ops) == 0:
+        return mesolve(H, state, tlist, e_ops=e_ops, args=args,
+                       options=options)
+
+    if isinstance(ntraj, list):
+        raise TypeError("No longer supported, use `result.expect_traj_avg`"
+                        "with the options `keep_runs_results=True`.")
+
+    mc = McSolver(H, c_ops, options=options)
+    result = mc.run(state, tlist=tlist, ntraj=ntraj, e_ops=e_ops,
+                    seed=seeds, target_tol=target_tol, timeout=timeout)
+    return result
+
+
+class _McTrajectorySolver(_TrajectorySolver):
+    """
+    Solver for one mcsolve trajectory. Created by a :class:`McSolver`.
+    """
+    name = "mcsolve"
+    _avail_integrators = {}
+
+    def __init__(self, rhs, c_ops, n_ops, *, options=None, _prepare=False):
+        self._c_ops = c_ops
+        self._n_ops = n_ops
+        super().__init__(rhs, options=options, _prepare=_prepare)
+
+    def _run(self, seed, state, tlist, e_ops):
+        self.collapses = []
+        self.generator = self.get_generator(seed)
+        self.target_norm = self.generator.random()
+        result = super()._run(self.generator, state, tlist, e_ops)
+        result.collapse = list(self.collapses)
+        result.seed = seed
+        return result
+
+    def start(self, state, t0, seed=None, *, safe_ODE=None):
+        self.collapses = []
+        self.generator = self.get_generator(seed)
+        self.target_norm = self.generator.random()
+        super().start(state, t0, seed=self.generator, safe_ODE=safe_ODE)
+
+    def prob_func(self, state):
+        if self.rhs.issuper:
+            return _data.norm.trace(unstack_columns(state))
+        return _data.norm.l2(state)**2
+
+    def norm_func(self, state):
+        if self.rhs.issuper:
+            return _data.norm.trace(unstack_columns(state))
+        return _data.norm.l2(state)
+
+    def _step(self, t, copy=True):
```

Comment:
The copy seems to be called for safe_ODE. It seems suspicious though to see no use of it in this function. all copy statements are set to False. I am not sure how this affect multiprocessing though.

Reply:
The returned state is always a copy because of the re-normalization at the end. In stochastic, it could be the array stored in the solver if `copy=False`.
Without copy here, either it always makes a copy, where stochastic fill suffer, or the function calling this will need to do one, then copies of copies will be made here.